src/commonMain/kotlin/org/kosat/CDCL.kt

```diff
@@ -48,96 +92,101 @@ class CDCL(val clauses: MutableList<MutableList<Int>>, initNumber: Int = 0) {
-        var clause: Int,
-        var level: Int,
-    )
+    private fun variable(lit: Int): Int = abs(lit)
 
-    // convert values to a possible satisfying result: if a variable less than 0 it's FALSE, otherwise it's TRUE
-    private fun variableValues(): List<Int> {
-        recoverAnswer()
-        return vars
-            .mapIndexed { index, v ->
-                when (v.status) {
-                    VarStatus.TRUE -> index
-                    VarStatus.FALSE -> -index
-                    else -> index
-                }
-            }.sortedBy { litIndex(it) }.filter { litIndex(it) > 0 }
+    // TODO rename
+    private fun watchedPos(lit: Int): Int {
+        return if (lit < 0) {
+            2 * (-lit)
+        } else {
+            2 * lit - 1
+        }
     }
 
-    // recover answer in terms of initial variables
-    private fun recoverAnswer() {
-        // updating vars for bve
-        val oldStatus = oldNumeration.mapIndexed { ind, _ -> vars[ind].status }
-        for (ind in 1..varsNumber) {
-            vars[ind].status = VarStatus.UNDEFINED
-        }
-        for (ind in 1..varsNumber) {
-            vars[oldNumeration[ind]].status = oldStatus[ind]
-            if (vars[oldNumeration[ind]].status == VarStatus.UNDEFINED) {
-                vars[oldNumeration[ind]].status = VarStatus.TRUE
-            }
-        }
-        varsNumber += deletingOrder.size
-        for (ind in deletingOrder.reversed()) {
-            var allTrue = true
-            for (clause in startOccurrence[ind]) {
-                var isTrue = false
-                for (lit in startClauses[clause]) {
-                    if (lit == ind) {
-                        continue
-                    }
-                    if (getStatus(lit) != VarStatus.FALSE) {
-                        isTrue = true
-                        break
-                    }
-                }
-                if (!isTrue) {
-                    allTrue = false
-                    break
-                }
-            }
-            if (allTrue) {
-                vars[ind].status = VarStatus.FALSE
-            } else {
-                vars[ind].status = VarStatus.TRUE
-            }
+    /** Interface **/
+
+    constructor() : this(mutableListOf<Clause>())
+
+    constructor(
+        initClauses: MutableList<Clause>,
+        initVarsNumber: Int = 0,
+        solverType: SolverType = SolverType.INCREMENTAL
+    ) : this(solverType) {
+        while (varsNumber < initVarsNumber) {
+            addVariable()
         }
+        initClauses.forEach { newClause(it) }
+        polarity = MutableList(varsNumber + 1) { VarStatus.UNDEFINED } // TODO is phaseSaving adapted for incremental?
     }
 
-    // values of variables
-    private val vars: MutableList<VarState> = MutableList(varsNumber + 1) { VarState(VarStatus.UNDEFINED, -1, -1) }
+    // public function for adding new variables
+    override fun addVariable() { // TODO simple checks of duplicate variables in newClause
```

Comment:
Either this method should be private, or it should return the newly allocated variable.